crawl-ref/source/traps.cc

```diff
@@ -1368,46 +1368,66 @@ void roll_trap_effects()
 }
 
 /***
- * Separate from the previous function so the trap triggers when crawl is in an
+ * Separate from roll_trap_effects so the trap triggers when crawl is in an
  * appropriate state
  */
 void do_trap_effects()
 {
-    const level_id place = level_id::current();
-    const Branch &branch = branches[place.branch];
-
     // Try to shaft, teleport, or alarm the player.
-    int roll = random2(3);
-    switch (roll)
+
+    // We figure out which possibilites are allowed before picking which happens
+    // so that the overall chance of being trapped doesn't depend on which
+    // possibilities are allowed.
+
+    // Don't shaft the player when shafts aren't allowed in the location or when
+    //  it would be into a dangerous end.
+    bool shaft_allowed = is_valid_shaft_effect_level();
+    // No alarms on the first 3 floors
+    bool alarm_allowed = env.absdepth0 > 3;
+    // Teleport effects are allowed everywhere, no need to check
+
+    trap_type trap_effect = TRAP_UNASSIGNED;
+
+    if (alarm_allowed)
     {
-        case 0:
-            // Don't shaft the player when we can't, and also when it would be into a
-            // dangerous end.
-            if (is_valid_shaft_level()
-               && !(branch.branch_flags & brflag::dangerous_end
-                    && brdepth[place.branch] - place.depth == 1))
-            {
-                dprf("Attempting to shaft player.");
-                you.do_shaft();
-            }
+        if (shaft_allowed)
+            trap_effect = random_choose(TRAP_ALARM, TRAP_SHAFT, TRAP_TELEPORT);
+        else
+            trap_effect = random_choose(TRAP_ALARM, TRAP_TELEPORT);
+    }
+    else
+    {
+        if (shaft_allowed)
+            trap_effect = random_choose(TRAP_SHAFT, TRAP_TELEPORT);
+        else
+            trap_effect = TRAP_TELEPORT;
+    }
```

Comment:
This can be simplified by making a vector of available trap effects, pushing back traps and teleports and then using the `random_iter` template in random.h

Reply:
Ah, sorry for letting that last change sit for so long, and thanks for wrapping it up for me!

Reply:
No problem!